File: apps/grokking-ml-kotlin-lib/src/main/kotlin/lib/Nb17BuildingDatasets.kt
package lib

import kotlin.random.Random
import org.jetbrains.kotlinx.dataframe.DataFrame
import org.jetbrains.kotlinx.dataframe.api.convertTo
import org.jetbrains.kotlinx.dataframe.api.getColumn
import org.jetbrains.kotlinx.dataframe.io.readCSV

/**
 * 原著ノートブック #17 `Chapter_11_Support_Vector_Machines/Building_the_datasets.ipynb`。
 *
 * 第 11 章（SVM）で使う 3 つのデータセットを **乱数で作る** 回である。
 *
 * 原著は種を固定していないので、**生成そのものは再現できない**。
 * できるのは 2 つ。
 *
 * 1. 同じ規則で生成器を書き、規則どおりの点が出ることを確かめる
 * 2. 原著が生成して配布した CSV を読み、**その規則で作られたことを検証する**
 *
 * 2 つ目が本命である。
 */
object Nb17BuildingDatasets {

    /** 座標の範囲。原著の `6 * random.random() - 3` */
    const val COORD_SCALE = 6.0
    const val COORD_OFFSET = -3.0

    /** 2 次元の点とラベル */
    data class Point(val x1: Double, val x2: Double, val y: Int)

    /** 1 つのデータセットの作り方 */
    data class Spec(
        val name: String,
        /** 規則どおりに作る点の数 */
        val points: Int,
        /** ラベルを乱数にする点（ノイズ）の数 */
        val noise: Int,
        /** ラベルを決める規則 */
        val rule: (Double, Double) -> Int,
    ) {
        val total: Int get() = points + noise
    }

    /** 直線 `x + y = 0.5` の上側なら 1 */
    fun linearRule(x: Double, y: Double): Int = if (x + y > 0.5) 1 else 0

    /** 原点を中心とする半径 √2.8 の円の内側なら 1 */
    fun oneCircleRule(x: Double, y: Double): Int = if (x * x + y * y < 2.8) 1 else 0

    /** (1, 0) と (-1, 0) を中心とする 2 つの円の **どちらか** の内側なら 1 */
    fun twoCirclesRule(x: Double, y: Double): Int {
        val left = (x - 1) * (x - 1) + y * y < 2
        val right = (x + 1) * (x + 1) + y * y < 2
        return if (left || right) 1 else 0
    }

    val LINEAR = Spec("linear", 50, 10, ::linearRule)
    val ONE_CIRCLE = Spec("one_circle", 100, 10, ::oneCircleRule)
    val TWO_CIRCLES = Spec("two_circles", 200, 20, ::twoCirclesRule)

    val SPECS = listOf(LINEAR, ONE_CIRCLE, TWO_CIRCLES)

    /**
     * 原著と同じ手順でデータセットを作る。
     *
     * 原著は種を固定していないので実行のたびに違うものが出る。
     * ここでは `seed` を渡せるようにして、テストで扱えるようにした。
     * **原著が配布している CSV とは一致しない**（一致しようがない）。
     */
    fun generate(spec: Spec, seed: Int): List<Point> {
        val random = Random(seed)
        fun coordinate() = COORD_SCALE * random.nextDouble() + COORD_OFFSET

        val points = List(spec.points) {
            val x = coordinate()
            val y = coordinate()
            Point(x, y, spec.rule(x, y))
        }
        val noise = List(spec.noise) {
            val x = coordinate()
            val y = coordinate()
            Point(x, y, random.nextInt(0, 2))
        }
        return points + noise
    }

    /**
     * 原著が生成して配布している CSV を読む。
     *
     * 先頭に無名の添字列が付いているが、列名で取り出すので気にしなくてよい。
     */
    fun load(spec: Spec): List<Point> {
        val frame: DataFrame<*> = DataFrame.readCSV(Datasets.path("${spec.name}.csv").toFile())
        val x1 = frame.getColumn("x_1").convertTo<Double>().toList()
        val x2 = frame.getColumn("x_2").convertTo<Double>().toList()
        val y = frame.getColumn("y").convertTo<Int>().toList()
        return x1.indices.map { Point(x1[it], x2[it], y[it]) }
    }

    /**
     * 規則とラベルが食い違う行の添字。
     *
     * ノイズとして入れた点は規則と無関係にラベルを振っているので、
     * **約半分がここに現れる**。規則どおりに作った先頭の点は 1 つも現れない。
     */
    fun ruleViolations(spec: Spec, data: List<Point>): List<Int> =
        data.indices.filter { spec.rule(data[it].x1, data[it].x2) != data[it].y }
}

File: apps/grokking-ml-fsharp-lib/src/GrokkingMlLib/Nb17BuildingDatasets.fs
/// 原著ノートブック #17 `Chapter_11_Support_Vector_Machines/Building_the_datasets.ipynb`。
///
/// 第 11 章（SVM）で使う 3 つのデータセットを **乱数で作る** 回である。
///
/// 原著は種を固定していないので、**生成そのものは再現できない**。
/// できるのは 2 つ。
///
/// 1. 同じ規則で生成器を書き、規則どおりの点が出ることを確かめる
/// 2. 原著が生成して配布した CSV を読み、**その規則で作られたことを検証する**
///
/// 2 つ目が本命である。
module GrokkingMlLib.Nb17BuildingDatasets

open Deedle

/// 座標の範囲。原著の `6 * random.random() - 3`
[<Literal>]
let CoordScale = 6.0

[<Literal>]
let CoordOffset = -3.0

/// 2 次元の点とラベル
type Point = { X1: float; X2: float; Y: int }

/// 1 つのデータセットの作り方
type Spec =
    { Name: string
      /// 規則どおりに作る点の数
      Points: int
      /// ラベルを乱数にする点（ノイズ）の数
      Noise: int
      /// ラベルを決める規則
      Rule: float -> float -> int }

    member this.Total = this.Points + this.Noise

/// 直線 `x + y = 0.5` の上側なら 1
let linearRule (x: float) (y: float) = if x + y > 0.5 then 1 else 0

/// 原点を中心とする半径 √2.8 の円の内側なら 1
let oneCircleRule (x: float) (y: float) = if x * x + y * y < 2.8 then 1 else 0

/// (1, 0) と (-1, 0) を中心とする 2 つの円の **どちらか** の内側なら 1
let twoCirclesRule (x: float) (y: float) =
    let left = (x - 1.0) ** 2.0 + y * y < 2.0
    let right = (x + 1.0) ** 2.0 + y * y < 2.0
    if left || right then 1 else 0

let linear =
    { Name = "linear"
      Points = 50
      Noise = 10
      Rule = linearRule }

let oneCircle =
    { Name = "one_circle"
      Points = 100
      Noise = 10
      Rule = oneCircleRule }

let twoCircles =
    { Name = "two_circles"
      Points = 200
      Noise = 20
      Rule = twoCirclesRule }

let specs = [ linear; oneCircle; twoCircles ]

/// 原著と同じ手順でデータセットを作る。
///
/// 原著は種を固定していないので実行のたびに違うものが出る。
/// ここでは `seed` を渡せるようにして、テストで扱えるようにした。
/// **原著が配布している CSV とは一致しない**（一致しようがない）。
let generate (spec: Spec) (seed: int) =
    let random = System.Random(seed)
    let coordinate () = CoordScale * random.NextDouble() + CoordOffset

    let points =
        List.init spec.Points (fun _ ->
            let x = coordinate ()
            let y = coordinate ()
            { X1 = x; X2 = y; Y = spec.Rule x y })

    let noise =
        List.init spec.Noise (fun _ ->
            let x = coordinate ()
            let y = coordinate ()
            { X1 = x; X2 = y; Y = random.Next(0, 2) })

    points @ noise

/// 原著が生成して配布している CSV を読む
let load (spec: Spec) =
    let frame = Datasets.loadFrame $"{spec.Name}.csv"
    let x1 = frame.GetColumn<float>("x_1") |> Series.values |> Array.ofSeq
    let x2 = frame.GetColumn<float>("x_2") |> Series.values |> Array.ofSeq
    let y = frame.GetColumn<int>("y") |> Series.values |> Array.ofSeq
    List.init x1.Length (fun i -> { X1 = x1.[i]; X2 = x2.[i]; Y = y.[i] })

/// 規則とラベルが食い違う行の添字。
///
/// ノイズとして入れた点は規則と無関係にラベルを振っているので、
/// **約半分がここに現れる**。規則どおりに作った先頭の点は 1 つも現れない。
let ruleViolations (spec: Spec) (data: Point list) =
    data
    |> List.indexed
    |> List.filter (fun (_, point) -> spec.Rule point.X1 point.X2 <> point.Y)
    |> List.map fst
